radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/EpochProofVerifierV2.java

```diff
@@ -18,83 +18,62 @@
 
 package com.radixdlt.statecomputer;
 
-import com.google.common.collect.Sets;
+import com.radixdlt.application.system.NextValidatorSetEvent;
+import com.radixdlt.application.system.state.EpochData;
 import com.radixdlt.consensus.bft.BFTNode;
-import com.radixdlt.consensus.epoch.EpochView;
+import com.radixdlt.consensus.bft.BFTValidator;
+import com.radixdlt.consensus.bft.BFTValidatorSet;
+import com.radixdlt.constraintmachine.REProcessedTxn;
 import com.radixdlt.engine.BatchVerifier;
 import com.radixdlt.engine.MetadataException;
-import com.radixdlt.ledger.ByzantineQuorumException;
 
-import java.util.Objects;
+import java.util.List;
 import java.util.stream.Collectors;
 
 public class EpochProofVerifierV2 implements BatchVerifier<LedgerAndBFTProof> {
 	@Override
-	public BatchVerifier.PerStateChangeVerifier<LedgerAndBFTProof> newVerifier(BatchVerifier.ComputedState computedState) {
-		return new PerEpochVerifier(computedState);
-	}
-
-	private final class PerEpochVerifier implements BatchVerifier.PerStateChangeVerifier<LedgerAndBFTProof> {
-		private final EpochView epochView;
-		private boolean epochChangeFlag = false;
-
-		private PerEpochVerifier(BatchVerifier.ComputedState initState) {
-			this.epochView = initState.get(EpochView.class);
-		}
+	public void testMetadata(LedgerAndBFTProof metadata, List<REProcessedTxn> txns) throws MetadataException {
+		NextValidatorSetEvent nextValidatorSetEvent = null;
+		for (int i = 0; i < txns.size(); i++) {
+			var processed = txns.get(i);
+			var nextEpochEvents = processed.getEvents().stream()
+				.filter(NextValidatorSetEvent.class::isInstance)
+				.map(NextValidatorSetEvent.class::cast)
+				.collect(Collectors.toList());
 
-		@Override
-		public void test(BatchVerifier.ComputedState computedState) {
-			if (epochChangeFlag) {
-				throw new ByzantineQuorumException("Additional commands added to end of epoch.");
-			}
-
-			var nextEpochView = computedState.get(EpochView.class);
-			if (nextEpochView.getEpoch() > epochView.getEpoch()) {
-				epochChangeFlag = true;
-			}
-		}
-
-		@Override
-		public void testMetadata(LedgerAndBFTProof metadata, BatchVerifier.ComputedState computedState) throws MetadataException {
-			// Verify that output of radix engine and signed output match
-			// TODO: Always follow radix engine as its a deeper source of truth and just mark validator
-			// TODO: set as malicious (RPNV1-633)
-			if (epochChangeFlag) {
-				if (metadata == null) {
-					throw new IllegalStateException();
+			if (!nextEpochEvents.isEmpty()) {
+				// TODO: Move this check into Meter
+				if (i != txns.size() - 1) {
+					throw new MetadataException("Additional txns added to end of epoch.");
```

Comment:
Do you think it's worth to have something like block-level validation?
- The max signature per block
- The max number of epoch update per block
- Epoch update should be the last state update

Moving everything to meters doesn't sound like the right direction

Reply:
We don't have a concept of blocks though. We will need metering in the long run though to have limits on computation per "block".

I don't think the idea of metering against a validatorSet is a bad concept as it reinforces the idea that we must protect against malicious quorums/cartels.